hkmc2/shared/src/main/scala/hkmc2/codegen/Block.scala

```diff
@@ -88,75 +88,74 @@ sealed abstract class Block extends Product with AutoLocated:
-              case None =>
-                val (rstRes, rstDefns) = rec(rest, armsDefns)
-                (Match(scrut, armsRes, None, rstRes), rstDefns)
-
-              case Some(dflt) =>
-                val (dfltRes, dfltDefns) = rec(dflt, armsDefns)
-                val (rstRes, rstDefns) = rec(rest, dfltDefns)
-                (Match(scrut, armsRes, S(dfltRes), rstRes), rstDefns)
-          
-        case Return(res, implct) => (b, acc)
-        case Throw(exc) => (b, acc)
-        case Label(label, body, rest) =>
-          if outerOnly then
-            val (rstRes, rstDefns) = rec(rest, acc)
-            (Label(label, body, rstRes), rstDefns)
-          else
-            val (bodyRes, bodyDefns) = rec(body, acc)
-            val (rstRes, rstDefns) = rec(rest, bodyDefns)
-            (Label(label, bodyRes, rstRes), rstDefns)
-        case Break(label) => (b, acc)
-        case Continue(label) => (b, acc)
-        case Begin(sub, rest) => 
-          val (subRes, subDefns) = rec(sub, acc)
-          val (rstRes, rstDefns) = rec(rest, subDefns)
-          (Begin(subRes, rstRes), rstDefns)
-        case TryBlock(sub, finallyDo, rest) =>
-          if outerOnly then
-            val (rstRes, rstDefns) = rec(rest, acc)
-            (TryBlock(sub, finallyDo, rstRes), rstDefns)
-          else
-            val (subRes, subDefns) = rec(sub, acc)
-            val (finallyRes, finallyDefns) = rec(rest, subDefns)
-            val (rstRes, rstDefns) = rec(rest, finallyDefns)
-            (TryBlock(subRes, finallyRes, rstRes), rstDefns)
-        case Assign(lhs, rhs, rest) => 
-          val (rstRes, rstDefns) = rec(rest, acc)
-          (Assign(lhs, rhs, rstRes), rstDefns)
-        case a @ AssignField(path, nme, result, rest) =>
-          val (rstRes, rstDefns) = rec(rest, acc)
-          (AssignField(path, nme, result, rstRes)(a.symbol), rstDefns)
-        case Define(defn, rest) => defn match
-          case ValDefn(owner, k, sym, rhs) => 
-            val (rstRes, rstDefns) = rec(rest, acc)
-            (Define(defn, rstRes), rstDefns)
-          case _ =>
-            val (rstRes, rstDefns) = rec(rest, defn :: acc)
-            (rstRes, rstDefns)
-        case HandleBlock(lhs, res, par, cls, handlers, body, rest) =>
+  
+  lazy val subBlocks: Ls[Block] = this match
+    case Match(_, arms, dflt, rest) => arms.map(_._2) ++ dflt.toList :+ rest
+    case Begin(sub, rest) => sub :: rest :: Nil
+    case TryBlock(sub, finallyDo, rest) => sub :: finallyDo :: rest :: Nil
+    case Assign(_, rhs, rest) => rest :: Nil
+    case AssignField(_, _, rhs, rest) => rest :: Nil
+    case Define(_, rest) => rest :: Nil
+    case HandleBlock(_, _, _, _, handlers, body, rest) => handlers.map(_.body) :+ body :+ rest
+    case _: Return | _: Throw | _: Label | _: Break | _: Continue | _: End | _: HandleBlockReturn => Nil
+  
+  // Moves definitions in a block to the top. Only scans one definition deep, i.e. definitions inside other definitions
```

Comment:
I wouldn't call this "one definition deep" (which would imply you look into at most one definition).

More like

```suggestion
  // Moves definitions in a block to the top. Only scans the top-level definitions of the block; i.e., definitions inside other definitions
```